eclipse-theia/theia
# Tab API implementation

#### What it does
resolves #11692

this implements the following parts of the vscode extension tab api: 

[Tab](https://code.visualstudio.com/api/references/vscode-api#Tab)
[TabChangeEvent](https://code.visualstudio.com/api/references/vscode-api#TabChangeEvent)
[TabGroup](https://code.visualstudio.com/api/references/vscode-api#TabGroup)
[TabGroupChangeEvent](https://code.visualstudio.com/api/references/vscode-api#TabGroupChangeEvent)
[TabGroups](https://code.visualstudio.com/api/references/vscode-api#TabGroups)
[TabInputCustom](https://code.visualstudio.com/api/references/vscode-api#TabInputCustom)
[TabInputTerminal](https://code.visualstudio.com/api/references/vscode-api#TabInputTerminal)
[TabInputText](https://code.visualstudio.com/api/references/vscode-api#TabInputText)
[TabInputTextDiff](https://code.visualstudio.com/api/references/vscode-api#TabInputTextDiff)
[TabInputWebview](https://code.visualstudio.com/api/references/vscode-api#TabInputWebview)

Following InputTypes are implemented but not used by the actual Events since there is no support for it currently in theia.
[TabInputNotebook](https://code.visualstudio.com/api/references/vscode-api#TabInputNotebook)
[TabInputNotebookDiff](https://code.visualstudio.com/api/references/vscode-api#TabInputNotebookDiff)

i sadly couldn't quite figure out what the preserve focus in the close functions is for. If someone knows please let me know.

#### How to test
you can find a small test-extension here: [tab-test-0.0.1.zip](https://github.com/eclipse-theia/theia/files/10500730/tab-test-0.0.1.zip)
or build it here: https://github.com/jonah-iden/tab-test

There is 3 main functionalities to test:
The listeners can partly be tested while testing the other functionalities

**test close current tab**
1.  open a tab
2. execute command "Close Active Tab"
3.  see current Tab should close. 
4. Also a message should pop up signifiying that a tabClose (or tab Group close if it was the last tab) event has occured

**test close current Group**
1. open multiple tabs. At best in a second group (but works also when there is only one group)
2.  execute command "Close Active TabGroup"
3. All tabs in current group should have been closed
4. messages for close events should also pop up

**other events to test**
- move tab in current group.
- activate Tab
- change tab from preview to normal
- pin tab



#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
